File: decimator-basic/decimator.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include "decimator.h"

void mixer_init(struct _mixer *mixer, double freq, unsigned int samplerate) {
    double phase_radian = 2*M_PI*freq/samplerate;
    mixer->osc = 1. + 0.j;
    mixer->phase = cos(phase_radian) + sin(phase_radian)*1.j;
}

void mixer(complex float *sampleout, const complex float *samplein, struct _mixer *mixer, int samplelen) {
    complex double       osc   = mixer->osc;             // register -> faster
    const complex double phase = mixer->phase;           // register -> faster

    for (int i=0; i<samplelen; i++) {
        sampleout[i] = samplein[i] * (complex float)osc; // mixing
        osc *= phase;                                    // next oscillator state
    }

    mixer->osc = osc;                                    // store back from register
}

// transition_bw: 0.05
static void calc_coeff(struct _decimstate *ds, float cutoff_rate, float transition_bw, enum Window window) {
    int taps_len=((int)(4.0/transition_bw + 0.5)) | 1; // odd
    int middle=taps_len/2;

    float *firtaps = ds->coeff_noncomplex;
    ds->coefflen = taps_len;

    float sum=0.;
    float wfuncval;
    for (int i=1; i<= middle; i++) {
        float rate = (float)i / middle / 2. + 0.5;

        switch (window) {
            case Blackman:
                wfuncval = 0.42-0.5*cos(2.*M_PI*rate) + 0.08*cos(4.*M_PI*rate);
                break;
            case Hamming:
                wfuncval = 0.54-0.46*cos(2.*M_PI*rate);
                break;
            default:
                puts("Bad filter type.");
                exit(-1);
                break;
        }

        firtaps[middle+i] = firtaps[middle-i] = sin(2.*M_PI*cutoff_rate*i) / i * wfuncval;
        sum += firtaps[middle-i] + firtaps[middle+i];
    }
    for (int i=0; i<taps_len; i++) {
        firtaps[i] /= sum;
    }
    fprintf(stderr, "taps_len: %d\n", taps_len);
}

// transition_bw: 0.05
int decimator_init(struct _decimstate *ds, unsigned int mixfreq, unsigned int samplerate, unsigned int decimfactor,
                       float transition_bw, enum Window window) {

    if (mixfreq != 0) { // mixer
        ds->is_mix = 1;
        mixer_init(&ds->mixer, mixfreq, samplerate);
    } else {
        ds->is_mix = 0;
    }

    ds->samplelen = 0;
    ds->decimfactor = decimfactor;
    calc_coeff(ds, 1./decimfactor, transition_bw, window);

    fprintf(stderr, "coefflen4: %d * 4\n", ds->coefflen);
    return 0;
}

int decimator(complex float *sampleout, const complex float *samplein, struct _decimstate *ds, int sampleinlen) {
    // mixcopy or simple copy?
    if (ds->is_mix) { // mix
        mixer(&ds->sample[ds->samplelen], samplein, &ds->mixer, sampleinlen);
    } else { // not mit --> memcopy
        memcpy(&ds->sample[ds->samplelen], samplein, &samplein[sampleinlen]-samplein);
    }
    ds->samplelen += sampleinlen;
    // decimate
    int j, k=0;
    for (j=0; j <= ds->samplelen - ds->coefflen - ds->decimfactor; j+= ds->decimfactor) {
        complex float tmp = 0.;

        for (int i=0; i < ds->coefflen; i++) {
            //tmp += ds->sample[i + j] * ds->coeff[i]
            tmp += ds->sample[i + j] * ds->coeff_noncomplex[i];
        }
        sampleout[k++] = tmp;

    }
    int pos = j;
    memmove(ds->sample, &ds->sample[pos], &ds->sample[ds->samplelen] - &ds->sample[pos]);
    ds->samplelen-=pos;
    return k; // return outlen
}

File: decimator-rust-basic/src/decimator.rs
#![allow(dead_code)]
use num_complex::Complex;
use std::f32::consts;

pub enum Window {
    Boxcar,
    Blackman,
    Hamming,
}

// transition_bw: 0.05
pub fn calc_coeff(cutoff_rate: f32, transition_bw: f32, window: Window) -> Vec<f32> {
    let taps_len=((4.0/transition_bw + 0.5) as usize) | 1; // odd
    let middle=taps_len/2;

    let mut sum=0.;
    let mut firtaps: Vec<f32> = vec![0.; taps_len];

    for i in 1..=middle {
        let rate=i as f32 / middle as f32 / 2. + 0.5;

        let wfuncval = match window {
            Window::Blackman =>
            // Explanation at Chapter 16 of dspguide.com, page 2
            // Blackman window has better stopband attentuation and passband ripple than Hamming, but it has slower rolloff.
                0.42-0.5*(2.*consts::PI*rate).cos() + 0.08*(4.*consts::PI*rate).cos(),
            Window::Hamming =>
                // Explanation at Chapter 16 of dspguide.com, page 2
                // Hamming window has worse stopband attentuation and passband ripple than Blackman, but it has faster rolloff.
                0.54-0.46*(2.*consts::PI*rate).cos(),
            _ => {
                println!("Bad type of filter! Exit.");
                std::process::exit(-1);
            },
        };

        firtaps[middle+i] = ((2.*consts::PI*cutoff_rate*i as f32).sin() / i as f32) * wfuncval;
        firtaps[middle-i] = firtaps[middle+i];
        sum += firtaps[middle-i] + firtaps[middle+i];
    }
    for f in &mut firtaps {
        *f /= sum;
    }

    firtaps
}

// ---------------------------------------------------------------

pub struct Decimator {
    samplerate: u32,
    samplefirbuf: Vec<Complex<f32>>,      // for fir func
    sampledecimbuf: Vec<Complex<f32>>,    // for decimate func
    sampledecimmixbuf: Vec<Complex<f32>>, // for decimate_mix func
    coeff: Vec<f32>,
    decimate: usize,
    is_mix: bool,
    oscillator: Complex<f32>,
    oscillator_phase: Complex<f32>,
}

impl Decimator {
    pub fn new(samplerate: u32, decimfactor: usize, transition_bw: f32, window: super::decimator::Window) -> Decimator {

        let cutoff_rate = 1./decimfactor as f32;
        let coeff = calc_coeff(cutoff_rate, transition_bw, window);

        Decimator {
             samplerate,
             samplefirbuf: Vec::new(),
             sampledecimbuf: Vec::new(),
             sampledecimmixbuf: Vec::new(),
             coeff,
             decimate: decimfactor,
             is_mix: false,
             oscillator: Complex::new(1., 0.),
             oscillator_phase: Complex::new(1., 0.),
        }
    }

    pub fn setfreq(&mut self, frequency: f64, phasereset: bool) {
        self.is_mix = frequency > 0.;
        let fpsr: f32 = (-2. * std::f64::consts::PI * frequency / f64::from(self.samplerate) ) as f32;
        self.oscillator_phase = Complex::new(fpsr.cos(), fpsr.sin());
        if phasereset {
            self.oscillator = Complex::new(1., 0.);
        }
    }

    // mixer & decimate
    pub fn decimator(&mut self, sample: &[Complex<f32>]) -> Vec<Complex<f32>> {
        let mut resvec = vec![];
        if self.decimate > self.coeff.len() { return resvec; }
        if self.is_mix {
            for s in sample {
                self.sampledecimmixbuf.push( s * self.oscillator );
                self.oscillator *= self.oscillator_phase;
            }
        } else {
            self.sampledecimmixbuf.extend(sample);
        }
        let mut pos = 0;
        for i in (0..self.sampledecimmixbuf.len()-self.coeff.len()).step_by(self.decimate) {
            resvec.push ( self.sampledecimmixbuf[i..i+self.coeff.len()].iter().zip(&self.coeff).map(|(sa, co)| sa * co).sum() );
            pos=i;
        }
        self.sampledecimmixbuf.drain(..pos+self.decimate);

        resvec
    }
}
